lib/routes/routeBackbeat.js

```diff
@@ -578,62 +619,86 @@ function putMetadata(request, response, bucketInfo, objMd, log, callback) {
         // To prevent this, the versionId field is only included in options when it is defined.
         if (versionId !== undefined) {
             options.versionId = versionId;
+            // In the MongoDB metadata backend, setting the versionId option leads to the creation
+            // or update of the version object, the master object is only updated if its versionId
+            // is the same as the version. This can lead to inconsistencies when replicating objects
+            // in the wrong order (can happen when retrying after failures), where the master object
+            // might point to a non current version. To prevent this, we need to set repairMaster to
+            // true which will update the master to the latest version available at the time of the put.
+            options.repairMaster = true;
         }
 
         // If the new null keys logic (S3C-7352) is not supported (compatibility mode), 'isNull' remains undefined.
         if (!nullVersionCompatMode) {
             options.isNull = isNull;
         }
 
-        log.trace('putting object version', {
-            objectKey: request.objectKey, omVal, options });
-        return metadata.putObjectMD(bucketName, objectKey, omVal, options, log,
-            (err, md) => {
+        return async.series([
+            // Zenko's Backbeat delegates replacing the account
+            // information to the destination's Cloudserver, as
+            // Vault admin APIs are not exposed externally.
+            next => getCanonicalIdsByAccountId(request.query?.accountId, log, (err, res) => {
```

Comment:
similarly: updating the account Id is not needed in case the request is coming from a remote CRR, and useless when used internally → should not degrade performance in that case.

if there is no better way, we could add an extra option on the query, defaulting to false and set to true only by CRR, to request this mapping?

Reply:
> updating the account Id is not needed in case the request is coming from a remote CRR, and useless when used internally → should not degrade performance in that case.

The `accountId` is a new parameter only passed by CRR, so this shouldn't affect performance of this API in other features.

I don't think there's a need for an extra parameter here, `accountId`'s only purpose is this, so CRR should just not pass it in S3C (or pass it if we decide to update both Cloudserver and Backbeat to the 9.0 and abandon the part where it updates the account info).

Reply:
so we need to condition the call to `getCanonicalIdsByAccountId`, and do it only if `accountId` is specified?

Reply:
The condition was already there inside `getCanonicalIdsByAccountId`, but looking at your other comment, sure i'll move it a level above.